ADR for phasing out legacy @ConfigRoot config classes support (#46196)
quarkusio/quarkus: adr/0008-phasing-out-config-classes.adoc

- We reduce our startup footprint (with marginal improvements in terms of RSS and startup time).

== Potential difficulties

=== Transitioning

Transitioning to config interfaces is a well-known and straightforward process.  For example, https://github.com/quarkusio/quarkus/pull/46070/files[this PR] illustrates such a transition in a non-simple extension (the MongoDB client extension).
It can be a bit tedious, though, for extensions with massive configurations or projects with a lot of extensions (Camel Quarkus comes to mind).

The main difficulty we identified is that the configuration is mutated in some extensions. It was possible with the legacy approach but never recommended. The new approach enforces the behavior, as the configuration is now immutable.
It should concern very few extensions. Also, we need to stop this harmful practice anyway.

=== Ecosystem

The Quarkus extension ecosystem is broad and diverse: some extensions are very well maintained and following the progress of Quarkus, some are lagging and adjusted when time permits.

If we phase out the legacy approach, we need the extension ecosystem ready.

> Indeed, each extension must be updated and released, as the legacy mechanism will be sunsetted.

Reviewer: ```suggestion
Indeed, each extension that still uses @ConfigRoot config classes must be updated and released, as the legacy mechanism will be sunsetted.
```

> For Quarkiverse extensions, the Quarkus community can be proactive as we can push changes there if needed.

Reviewer: didn't we submit prs for those we could find?
How about camel and others in platform ? 

if we have them submitted lets link to them.

Author: I can adjust the wording as I adjusted most if not all of them after I wrote this.

But I see no point in pointing to all the PRs in the ADR.

= Phasing out legacy config classes

* Status: proposed
* Date: 2025-02-06
* Authors: @gsmet @radcortez

== Context and Problem Statement

We currently have two competing implementations to handle configuration properties in Quarkus extensions:

- The legacy config class approach: it uses **classes** annotated with `@io.quarkus.runtime.annotations.ConfigRoot` and **fields** annotated with `@io.quarkus.runtime.annotations.ConfigItem` for the properties.
- The new config interface approach: it uses **interfaces** annotated with `@io.smallrye.config.ConfigMapping` and `@io.quarkus.runtime.annotations.ConfigRoot` (for the config phase) and **methods** for the properties.

> The latter approach is what we have been recommending for quite some time now (it actually requires a special flag for the extension annotation processor to use the legacy config class approach).

Author: For instance in the Quarkiverse: https://github.com/search?q=org%3Aquarkiverse%20legacyConfigRoot&type=code .

It's very likely I will create PRs for most of them before end of week.

> The latter approach is what we have been recommending for quite some time now (it actually requires a special flag for the extension annotation processor to use the legacy config class approach).

Author: FWIW, I spent some time today to create PRs for all the Quarkiverse extensions that were using the `legacyConfigRoot` compiler argument.

- Classes were mutable, whereas Quarkus configuration is immutable, causing hard-to-debug problems and unexpected behaviours.
- It only worked for extensions and in the context of Quarkus.

SmallRye Config provides the currently recommended mechanism which can be used in any context (even without Quarkus). Using an interface instead of class makes it clear that the Config cannot be changed.
On top of that, many of the new features were more straightforward to implement in this new model.

There has been a massive effort to migrate from the former approach to the latter and we are happy to announce we migrated all the Core configuration to the `@ConfigMapping` approach.
There are a few legacy classes left but they are there for compatibility purposes, for some config properties massively consumed outside of Quarkus that we didn't want to break (yet!).

At this stage, it is time to start planning phasing out the legacy approach so that:

- We avoid maintaining this code forever. It is complex and the bandwidth of the person maintaining it is limited.
- We avoid some bugs that were never fixed in the config class approach due to either limited bandwidth or high complexity.

> - We reduce our startup footprint.

Author: I would expect very marginal improvements in both (less class loaded, a bit less work).

> By phasing out, we mean that the code won't be there anymore and that the approach with `@ConfigRoot` annotated **classes** won't be functional anymore, including for external extensions.

Reviewer 2: ```suggestion
By phasing out, we mean that the approach with `@ConfigRoot` annotated **classes** will no longer be functional, including for external extensions.
```

> The good thing is by migrating the Core configuration, we have loads of examples we can point people to and we are also very confident in the solidity of the `@ConfigMapping` code in SmallRye Config.

Reviewer 2: ```suggestion
With the Quarkus core extension having switched to the new model, we have many examples we can point people to, and we are also very confident in the robustness of the `@ConfigMapping` support from SmallRye Config.
```

> Roberto has also led an effort to improve the performances of the new approach and allow it to scale better and results are promising (and still improving).

Reviewer 2: ```suggestion
@radcortez  has also improved the performance of the new approach and allowed it to scale better.
```

Reviewer 2: I would move this in the thrid point above (about the footprint)

The main goal of this ADR is to propose an approach for the transition that reduces friction.

== Potential difficulties

=== Transitioning

> 

Author: See the reasons why we are doing this. The whole idea is to not to have to maintain two different implementations forever.

> Transitioning to config interfaces is a well known and straightforward process.

Reviewer 2: ```suggestion
Transitioning to config interfaces is a well-known and straightforward process.  For example, [this PR](https://github.com/quarkusio/quarkus/pull/46070/files) illustrates such a transition in a non-simple extension (the MonboDB client extension).
```

> It can be a bit tedious though for extensions with massive configuration or projects with a lot of extensions (Camel Quarkus comes to mind).

Reviewer 2: ```suggestion
It can be a bit tedious, though, for extensions with massive configurations or projects with a lot of extensions (Camel Quarkus comes to mind).
```

> The main difficulty we identified is that in some extensions, we actually mutate the configuration, which was possible (but definitely not recommended!) for public fields and is not possible with the new interface/method approach.

Reviewer 2: ```suggestion
The main difficulty we identified is that the configuration is mutated in some extensions. It was possible with the legacy approach but never recommended. The new approach enforces the behavior, as the configuration is now immutable.
```

> It should concern very few extensions though and we need to stay away from this bad practice anyway.

Reviewer 2: ```suggestion
It should concern very few extensions. Also, we need to stop this harmful practice anyway.
```

=== Ecosystem

The Quarkus extension ecosystem is broad and diverse: some extensions are very well maintained and following the progress of Quarkus, some are lagging and adjusted when time permits.

> If we phase out the legacy approach, we need the extension ecosystem to be ready for it.

Reviewer 2: ```suggestion
If we phase out the legacy approach, we need the extension ecosystem ready.
Indeed, each extension must be updated and released, as the old mechanism will be sunsetted.
```

For Quarkiverse extensions, the Quarkus community can be proactive as we can push changes there if needed.

> Things become a bit more blurry for extensions hosted outside of the Quarkiverse for which we will need the maintainers to be doing the work.

Reviewer 2: ```suggestion
Things become a bit more complex for extensions hosted outside of the Quarkiverse, for which we will need the maintainers to do the work. We plan to contact the known maintainers and add notes in the release notes and migration guides.
```

=== Compatibility classes

During the migration, for a few key extensions, we decided to keep small and targeted config classes around to maintain the compatibility for all the extensions that consume this configuration (either as a CDI bean at runtime or during build time).
This is for instance the case of the HTTP configuration coming from Vert.x HTTP.

> When we phase out the legacy approach, these classes will have to go: we need to make sure the extension ecosystem has moved to the new interfaces (or to a proper API if we put one in place).

Reviewer 2: ```suggestion
When we phase out the legacy approach, these classes will be removed: we need to ensure the extension ecosystem has moved to the new interfaces (or to a proper API if we put one in place).
```

> == Proposed approach

Reviewer 2: ```suggestion
== Sunsetting plan
```

Given all the items presented above, the key for this transition to be a success is communication and planning.
Here is the approach we recommend:

> - Step 1 - Communicate first about where we introduced compatibility config classes and that people should consume the new config interface ASAP

Reviewer 2: ```suggestion
- **Step 1** - Communicate first about where we introduced compatibility config classes and that people should consume the new config interface ASAP
```

> - Step 2 - Communicate early about this upcoming transition

Reviewer 2: ```suggestion
- **Step 2** - Communicate early about this upcoming transition

```

  - Blog post + post on quarkus-dev
  - Post in the Quarkiverse community discussions
  - Prepare a guide on how to do it. Make sure we add useful pointers with example pull requests.
  - Make sure we provide points of contact in the community for people having difficulties (or who just don't have the time to do it and want someone else to handle it)
  - Make sure we identify difficulties with potential solutions
  - Announce specific versions and dates for the tentative phase out so that people can schedule their work

> - Step 3 - In version X.Y (with at least two months between previous communication and release), throw an error in the extension annotation processor when legacy `@ConfigRoot` classes are detected

Reviewer 2: ```suggestion
- **Step 3** - In version X.Y (with at least two months between previous communication and release), throw an error in the extension annotation processor when legacy `@ConfigRoot` classes are detected

```

  - It will allow us to detect the overall status in Ecosystem CI (at least for public extensions)
  - Maybe provide a way to forcefully override this error for one version, and drop it in the next version
  - Keep the support for legacy `@ConfigRoot` classes in quarkus-core though as we want the extensions to still work until they are updated
  - Announce it via blog post + post on quarkus-dev

> - Step 4 - In version X.(Y + 2), if we are satisfied with the state of the Ecosystem, drop the legacy `@ConfigRoot` classes support from Quarkus entirely

Reviewer 2: ```suggestion
- **Step 4 **- In version X.(Y + 2), if we are satisfied with the state of the Ecosystem, drop the legacy `@ConfigRoot` classes support from Quarkus entirely
```

Reviewer 2: including the compatibility classes.

== Proposed schedule

Given we were able to get everything ready for 3.19 and 3.20 LTS, here is the schedule we would suggest:

- Step 1 - For the 3.19 release on February 26th
- Step 2 - For the 3.19 release - and we should amplify the message when 3.20 LTS gets out
- Step 3 - For the 3.22 release on April 30th

> - Step 4 - For the 3.24 release on June 25th

Author: @maxandersen no, apps will run without a hicchup, provided the extensions have been converted to `@ConfigMapping`.

We already have a warning in the annotation processor when building with legacyConfigRoot. It has been there since I rewrote the annotation processor:

```
        if (!useConfigMapping) {
            processingEnv.getMessager().printMessage(Diagnostic.Kind.WARNING, "Extension module " + extensionModule.artifactId()
                    + " config implementation is deprecated. Please migrate to use @ConfigMapping: https://quarkus.io/guides/writing-extensions#configuration");
        }
```

> - Step 4 - For the 3.24 release on June 25th

Author: @radcortez would it be possible to also add a warning when building an application with the list of legacy config roots (ideally the extension name would be handy, but I'm not sure we can get the extension name and I wouldn't create an additional infrastructure just for that)?

> - Step 4 - For the 3.24 release on June 25th

Reviewer 3: Warning the legacy config roots is easy. I'm not sure about finding their extensions. We could list the FQN, which should give you a pretty good idea of where it is coming from.

> - Step 4 - For the 3.24 release on June 25th

Reviewer 3: https://github.com/quarkusio/quarkus/pull/46314

> - Step 4 - For the 3.24 release on June 25th

Reviewer 2: I would add a paragraph like:

This effort is essential. It:

- reduces the codebase complexity
- avoid ambiguities and bad practices
- improves startup performances

Author: Done!

> - Step 4 - For the 3.24 release on June 25th

Reviewer: `no, apps will run without a hicchup, provided the extensions have been converted to @ConfigMapping.`

that is what I mean - users app risk breaking and they have no ability to fix it unless all their extensions moved too.

That probably is fine especially if we can fix all known quarkiverse extensions at least.

But just want us to be clear in communication that Users will possibly have a breakage and they can't fix it in their code or config it *requires* extensions to be released.
